api/controllers/WebActionsController.js

```diff
@@ -38,27 +38,45 @@ module.exports = {
    * @return {object}
    */
   createMobileCommonsRequest: function(req) {
-    let optInPath;
-
-    if (req.body.partner) {
-      optInPath = PartnerService.getOptInPath(req.body.partner);      
-    } else {
-      optInPath = this.MOBILE_COMMONS_OPTIN
-    }
-    
-    return {
+    let data = {
       form: {
-        'opt_in_path[]': optInPath,
+        'opt_in_path[]': req.body.opt_in_path || this.MOBILE_COMMONS_OPTIN,
         'person[first_name]': req.body.first_name,
         'person[phone]': req.body.phone,
         'person[email]': req.body.email,
-        'person[send_gifs]': typeof req.body.send_gifs === 'undefined'
-          ? 'no'
-          : 'yes',
+        'person[send_gifs]': typeof req.body.send_gifs === 'undefined' ? 'no' : 'yes',
         'person[referral_code]': ReferralCodes.encode(req.body.phone),
         'person[date_signed_up]': new Date().toISOString(),
       },
     };
+
+    if (req.body.extras) {
+      const keys = Object.keys(req.body.extras);
+      for (const key of keys) {
+        data.form[`person[${key}]`] = req.body.extras[key];
+      }
+    }
+
+    // If an array of friends and opt-in path for them are provided, add it to
+    // the submission data.
+    if (req.body.friends_opt_in_path && req.body.friends.length > 0) {
+      data.form['friends_opt_in_path'] = req.body.friends_opt_in_path;
+
+      let friendCounter = 0;
+      for (const friend of req.body.friends) {
+        if (typeof friend === 'object' && friend.first_name && friend.phone) {
+          const code = ReferralCodes.encode(friend.phone);
+
+          data.form[`friends[${friendCounter}][phone]`] = friend.phone;
+          data.form[`friends[${friendCounter}][first_name]`] = friend.first_name;
+          data.form[`friends[${friendCounter}][referral_code]`] = code;
```

Comment:
The generated referral code won't be useful to check for fake numbers. We'll probably want to check directly against the phone numbers a user enters in. And we'll probably want to do the validation on the client-side for quicker feedback.

We could maybe use this library https://github.com/catamphetamine/libphonenumber-js which is based off a Google library. But it doesn't seem to prevent against xxx-555-01xx sort of numbers. And wouldn't help with other kinds of validation we could do like just making sure all 3 numbers aren't the same.